src/core/lib/gprpp/thd_posix.cc

```diff
@@ -0,0 +1,196 @@
+  Thread* thread;
+  void (*body)(void* arg); /* body of a thread */
+  void* arg;               /* argument to a thread */
+  const char* name;        /* name of thread. Can be nullptr. */
+};
+
+/*****************************************
+ * Only used when fork support is enabled
+ */
+
+void inc_thd_count() {
+  if (grpc_fork_support_enabled()) {
+    gpr_mu_lock(&g_mu);
+    g_thread_count++;
+    gpr_mu_unlock(&g_mu);
+  }
+}
+
+void dec_thd_count() {
+  if (grpc_fork_support_enabled()) {
+    gpr_mu_lock(&g_mu);
+    g_thread_count--;
+    if (g_awaiting_threads && g_thread_count == 0) {
+      gpr_cv_signal(&g_cv);
+    }
+    gpr_mu_unlock(&g_mu);
+  }
+}
+
+}  // namespace
+
+Thread::Thread(const char* thd_name, void (*thd_body)(void* arg), void* arg,
+               bool* success)
+    : real_(true), alive_(false), started_(false), joined_(false) {
+  gpr_mu_init(&mu_);
+  gpr_cv_init(&ready_);
+  pthread_attr_t attr;
+  /* don't use gpr_malloc as we may cause an infinite recursion with
+   * the profiling code */
+  thd_arg* a = static_cast<thd_arg*>(malloc(sizeof(*a)));
+  GPR_ASSERT(a != nullptr);
+  a->thread = this;
+  a->body = thd_body;
+  a->arg = arg;
+  a->name = thd_name;
+  inc_thd_count();
+
+  GPR_ASSERT(pthread_attr_init(&attr) == 0);
+  GPR_ASSERT(pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE) == 0);
+
+  pthread_t p;
+  alive_ = (pthread_create(&p, &attr,
+                           [](void* v) -> void* {
+                             thd_arg a = *static_cast<thd_arg*>(v);
+                             free(v);
+                             if (a.name != nullptr) {
+#if GPR_APPLE_PTHREAD_NAME
+                               /* Apple supports 64 characters, and will
+                                * truncate if it's longer. */
+                               pthread_setname_np(a.name);
+#elif GPR_LINUX_PTHREAD_NAME
+                               /* Linux supports 16 characters max, and will
+                                * error if it's longer. */
+                               char buf[16];
+                               size_t buf_len = GPR_ARRAY_SIZE(buf) - 1;
+                               strncpy(buf, a.name, buf_len);
+                               buf[buf_len] = '\0';
+                               pthread_setname_np(pthread_self(), buf);
+#endif  // GPR_APPLE_PTHREAD_NAME
+                             }
+
+                             gpr_mu_lock(&a.thread->mu_);
+                             if (!a.thread->started_) {
+                               gpr_cv_wait(&a.thread->ready_, &a.thread->mu_,
+                                           gpr_inf_future(GPR_CLOCK_MONOTONIC));
+                             }
+                             gpr_mu_unlock(&a.thread->mu_);
+
+                             (*a.body)(a.arg);
+                             dec_thd_count();
+                             return nullptr;
+                           },
+                           a) == 0);
```

Comment:
I missed seeing this `a` and got very confused :).  Would you be able to rename this variable to something to other than `a` so that it is easier to search for the variable and find it :)